src/Workspaces/Remote/Core/ExportProviderBuilder.cs

```diff
@@ -0,0 +1,220 @@
+    private static string GetCompositionCacheFilePath(string cacheDirectory, string catalogPrefix, ImmutableArray<string> assemblyPaths)
+    {
+        return Path.Combine(cacheDirectory, $"{catalogPrefix}.{ComputeAssemblyHash(assemblyPaths)}{CatalogSuffix}");
+
+        static string ComputeAssemblyHash(ImmutableArray<string> assemblyPaths)
+        {
+            // Ensure AssemblyPaths are always in the same order.
+            assemblyPaths = assemblyPaths.Sort();
+
+            var hashContents = new StringBuilder();
+
+            // This should vary based on .NET runtime major version so that as some of our processes switch between our target
+            // .NET version and the user's selected SDK runtime version (which may be newer), the MEF cache is kept isolated.
+            // This can be important when the MEF catalog records full assembly names such as "System.Runtime, 8.0.0.0" yet
+            // we might be running on .NET 7 or .NET 8, depending on the particular session and user settings.
+            hashContents.Append(Environment.Version.Major);
+
+            foreach (var assemblyPath in assemblyPaths)
+            {
+                // Include assembly path in the hash so that changes to the set of included
+                // assemblies cause the composition to be rebuilt.
+                hashContents.Append(assemblyPath);
+                // Include the last write time in the hash so that newer assemblies written
+                // to the same location cause the composition to be rebuilt.
+                hashContents.Append(File.GetLastWriteTimeUtc(assemblyPath).ToString("F"));
+            }
+
+            // Create base64 string of the hash.
+            var hashAsBase64String = Checksum.Create(hashContents.ToString()).ToString();
+
+            // Convert to filename safe base64 string.
+            return hashAsBase64String.Replace('+', '-').Replace('/', '_').TrimEnd('=');
+        }
+    }
+
+    private static async Task WriteCompositionCacheAsync(string compositionCacheFile, CompositionConfiguration config, bool performCleanup, Action<string> logError, CancellationToken cancellationToken)
+    {
+        try
+        {
+            await Task.Yield().ConfigureAwait(false);
+
+            if (Path.GetDirectoryName(compositionCacheFile) is string directory)
+            {
+                var directoryInfo = Directory.CreateDirectory(directory);
+
+                if (performCleanup)
+                {
+                    // Delete any existing cached files.
+                    foreach (var fileInfo in directoryInfo.EnumerateFiles($"*{CatalogSuffix}"))
+                    {
+                        fileInfo.Delete();
+                        cancellationToken.ThrowIfCancellationRequested();
+                    }
+                }
+            }
+
+            CachedComposition cachedComposition = new();
+            var tempFilePath = Path.Combine(Path.GetTempPath(), Path.GetTempFileName());
```

Comment:
This should probably create the temporary file in the same directory for a few reasons:

1. Security: on Windows %TEMP% is isolated from users; that's not necessarily the case on Linux. This at least creates a small information leak in that the user's installed extensions are present in the MEF cache.
2. Reliability: it makes a stronger guarantee that that File.Move() can't corrupt stuff. If it's on the same volume/filesystem, that becomes a bit of file system recordkeeping and that's it. And you can use File.Replace() if needed. If they're on different file/systems, then the move is under the cover copying bits, and a badly timed power outage means your cache file might get corrupted.

Reply:
makes sense